docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -158,7 +160,23 @@ Grantee: cosmos1ghi...jkl
 End of transaction messages
 ```
 
-### 9. Signing Payload and Wire Format
+### 10. Require signing over the `TxBody` and `AuthInfo` raw bytes
+
+Recall that the transaction bytes merklelized on chain are the Protobuf binary serialization of [TxRaw](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0/proto/cosmos/tx/v1beta1/tx.proto#L33), which contains the `body_bytes` and `auth_info_bytes`. Moreover, the transaction hash is defined as the SHA256 hash of the `TxRaw` bytes. We require that the user signs over these canonical bytes in SIGN_MODE_TEXTUAL, more specifically over the following string:
```

Comment:
I don't have a great alternative but "these canonical bytes" sounds misleading to me in two ways.
- The idea of `body_bytes` and `auth_info_bytes` is that any proto serialization is accepted and then fixed. So what exactly is "canonical"?
- You talk about TxRaw before, then about "these [...] bytes". But you do not refer to TxRaw since this includes the signatures.

Reply:
Canonical shouldn't be used here. These bytes are "canonical", they're just the bytes in `TxRaw` 
```suggestion
Recall that the transaction bytes merklelized on chain are the Protobuf binary serialization of [TxRaw](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0/proto/cosmos/tx/v1beta1/tx.proto#L33), which contains the `body_bytes` and `auth_info_bytes`. Moreover, the transaction hash is defined as the SHA256 hash of the `TxRaw` bytes. We require that the user signs over these bytes in SIGN_MODE_TEXTUAL, more specifically over the following string:
```

Reply:
Canonical bytes are the bytes of the submitted (i.e. unsigned) tx that in the block and end up on-chain. Whatever serialization is used in the block _is_ accepted and fixed as the truth. Alternative serializations may be equivalent semantically, but signatures don't verify their integrity.

Reply:
I never came across this usage of "canonical bytes" in Cosmos sign mode development. My understanding of canonical is the one of a canonical form. This canonical form is not defined and `body_bytes` and `auth_info_bytes` can contain an arbitrary protobuf serialization.

Reply:
This is kind of the point I'm trying to get across 😅 The way that Cosmos treats signatures when checking for integrity isn't correct. A cryptographic signature of some message verifies the integrity of _that literal message_, the specific bytes given to the signature verification function, and nothing more. 

```
type Msg struct { From string, To string, Amount int, Memo string }
func (m *Msg) RenderText { return Sprintf("From %s, To %s, Amount %d", m.From, m.To, m.Amount) }
func (m *Msg) RenderJSON { return json.Marshal(m) }
```

The text rendering of this message doesn't include the memo field. A signature on the RenderText bytes therefore does not validate the integrity of that information.

```
m1 = Msg{From: "me", To: "you", Amount: 123, Memo: "foo"}
s1text = Sign(m1.RenderText)
s2json = Sign(m1.RenderJSON)

m2 = Msg{From: "me", To: "you", Amount: 123, Memo: "bar"}
s2text = Sign(m2.RenderText)
s2json = Sign(m2.RenderJSON)
```

m1 and m2 are different transactions. It should not be possible to sign them in a way that the signatures are equivalent. But s1text and s2text are the same. This is a problem. It means that you can take a legitimate signed msg, change any data in it not captured by the signature, and the signature will valid. The signature therefore does not prove integrity. Neither can you treat a signature made in one mode as equivalent to a signature made in another mode. The rendering modes are free to include different subsets of information, in different encoded forms, from each other.

Reply:
This inline comment is probably not a great place to discuss Amino JSON or Textual sign mode transformations, so I'll not address this part.

The whole point of this PR is to fix a single byte representation, the one that is stored in the block. So yes, agreed in general. However, it's not a "canonical" serialzation as protobuf does not have this concept. It is any valid protobuf serialization. We hit this limitation many times and `body_bytes`/`auth_info_bytes` only exist because of it. You could argue that protobuf is a bad choice because its serialization is not unique and I would not disagree.

Another note here: It is not possible to sign the transaction written in the block because the transaction contains all signatures.

Reply:
It's fine if Protobuf doesn't serialize a tx/msg/etc. in a single way. But whatever serialization produces the bytes which end up in the block becomes, by definition, the canonical serialization for the corresponding value. And when I write "canonical" I don't mean from the perspective of Protobuf. Again, signatures don't validate parsed or transformed versions of a value, they validate the specific bytes passed to the signing function. When you sign a tx/msg/etc. you're signing the specific serialized form provided to the signing function. The signature is not transitive to the underlying value.

Reply:
@peterbourgon In none of the sign modes in the SDK do we sign over the on-chain (what you call canonical) tx bytes . If that's a concern, let's discuss this elsewhere:
- you can create an issue
- post on Discord
- on if it's security-sensitive, report a security bug.


---

Let's get back on topic: signing over a pinned byte representation of the tx. I removed the "wording".